pkg/pillar/base/logobjecttypes.go

```diff
@@ -0,0 +1,291 @@
+// Copyright (c) 2020 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package base
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/satori/go.uuid"
+	log "github.com/sirupsen/logrus"
+)
+
+// LogObjectType : Predefined log object types
+type LogObjectType string
+
+// The following is NOT an extensive list.
+// Intent is to create a reference list for engineers to get an idea of its usage.
+const (
+	UnknownLogType         LogObjectType = ""
+	LogType                LogObjectType = "log"
+	ImageLogType           LogObjectType = "image"
+	RelationLogType        LogObjectType = "relation"
+	AppInstanceLogType     LogObjectType = "app_instance"
+	NetworkInstanceLogType LogObjectType = "network_instance"
+)
+
+// LogObject : Holds all key value pairs to be logged later.
+type LogObject struct {
+	Initialized bool
+	Fields      map[string]interface{}
+}
+
+var logObjectMap = make(map[string]*LogObject)
+
+// LoggableObject :
+type LoggableObject interface {
+	Key() string
+	LogCreate()
+	LogModify(old interface{})
+	LogDelete()
+}
+
+// NewLogObject :
+// objType -> log, action, relation, app_instance, image etc
+// objName -> App instance name, name of file being downloaded etc
+// objUUID -> UUID of the object if present (App instance UUID) or UUID of parent object
+// keys -> Keys on which GetValue is called to get values
+// objType and objName are mandatory parameters
+func NewLogObject(objType LogObjectType, objName string, objUUID uuid.UUID, key string) *LogObject {
+	if objType == UnknownLogType || len(objName) == 0 {
+		return nil
+	}
+	// Check if we already have an object with the given key
+	object, ok := logObjectMap[key]
+	if !ok {
+		object = new(LogObject)
+	}
+	fields := make(map[string]interface{})
+	fields["obj_type"] = objType
+	fields["obj_name"] = objName
+	fields["obj_uuid"] = objUUID.String()
+	object.Initialized = true
+	object.Fields = fields
+	logObjectMap[key] = object
```

Comment:
If you found it in the map, why would you change it?

Reply:
And if you create a new you can use InitLogObject

Reply:
makes sense not to change an existing entry. Will use InitLogObject.